src/utils/SpatialTreeInterface/SpatialTreeInterface.jl

```diff
@@ -0,0 +1,226 @@
+"""
+isspatialtree(::T) where T = isspatialtree(T)
+isspatialtree(::Type{<: Any}) = false
+
+
+"""
+    getchild(node)
+
+Return an iterator over all the children of a node.
+This may be materialized if necessary or available,
+but can also be lazy (like a generator).
+"""
+getchild(node) = AbstractTrees.children(node)
+
+"""
+    getchild(node, i)
+
+Return the `i`-th child of a node.
+"""
+getchild(node, i) = getchild(node)[i]
+
+"""
+    nchild(node)
+
+Return the number of children of a node.
+"""
+nchild(node) = length(getchild(node))
+
+"""
+    isleaf(node)
+
+Return true if the node is a leaf node, i.e., there are no "children" below it.
+[`getchild`](@ref) should still work on leaf nodes, though, returning an iterator over the extents stored in the node - and similarly for `getnodes.`
+"""
+isleaf(node) = error("isleaf is not implemented for node type $(typeof(node))")
+
+"""
+    child_indices_extents(node)
+
+Return an iterator over the indices and extents of the children of a node.
+
+Each value of the iterator should take the form `(i, extent)`.
+"""
+function child_indices_extents(node)
+    return zip(1:nchild(node), getchild(node))
+end
+
+# ## Query functions
+# These are generic functions that work with any spatial tree type that implements the interface.
+
+
+"""
+    do_query(f, predicate, tree)
+
+Call `f(i)` for each index `i` in the tree that satisfies `predicate(extent(i))`.
+
+This is generic to anything that implements the SpatialTreeInterface, particularly the methods
+[`isleaf`](@ref), [`getchild`](@ref), and [`child_extents`](@ref).
+"""
+function do_query(f::F, predicate::P, node::N) where {F, P, N}
+    if isleaf(node)
+        for (i, leaf_geometry_extent) in child_indices_extents(node)
+            if predicate(leaf_geometry_extent)
+                @controlflow f(i)
+            end
+        end
+    else
+        for child in getchild(node)
+            if predicate(GI.extent(child))
+                @controlflow do_query(f, predicate, child)
+            end
+        end
+    end
+end
+
+function do_query(predicate, node)
+    a = Int[]
+    do_query(Base.Fix1(push!, a), predicate, node)
+    return a
+end
+
+
+"""
+    query(tree, predicate)
+
+Return a sorted list of indices of the tree that satisfy the predicate.
+"""
+function query(tree, predicate)
+    a = Int[]
+    do_query(Base.Fix1(push!, a), sanitize_predicate(predicate), tree)
+    return sort!(a)
+end
+
+
+"""
+    sanitize_predicate(pred)
+
+Convert a predicate to a function that returns a Boolean.
+
+If `pred` is an Extent, convert it to a function that returns a Boolean by intersecting with the extent.
+If `pred` is a geometry, convert it to an extent first, then wrap in Extents.intersects.
+
+Otherwise, return the predicate unchanged.
+
+
+Users and developers may overload this function to provide custom behaviour when something is passed in.
+"""
+sanitize_predicate(pred::P) where P = sanitize_predicate(GI.trait(pred), pred)
```

Comment:
```suggestion
sanitize_predicate(pred) = sanitize_predicate(GI.trait(pred), pred)
```

Not used

Reply:
IIRC this helped with inferrability which is why I put it in.

Reply:
Oh right it's a function. Yes keep it